### ISSUE TITLE
add ram limit to net for production

### ISSUE THREAD

**Issue by danFromTelAviv** [pytorch/pytorch#20396]
## 🚀 Feature
Allow a ram limitation that would slow down the inference in favor of limiting ram usage. 

## Motivation
In many production settings, we want to use very cheap servers that have specific limitations on the ram. However, we get all sorts of samples ranging in sizes (audio in my case, but this could be true for text or other variable length samples ). The default way to deal with this is to split the samples into overlaping parts. run the net and stitch the results. I have no doubt that there is a better way to process rnns and cnns ( and maybe transformers ) in such a way where the ram usage is limited to the required ram of one ( or "n") timestep per layer.

## key features 
1) allow to set a max ram - throw an error if its lower than possible.
2) optimize any sequence of operations such as stacked rnns and cnns to limit ram at the cost of runtime. 

I think this will help many users in serving models for production.

**Comment by soumith** [comment]
If you just need to run inference, running the inference in a `with torch.no_grad():` block will store no intermediate buffers at all, and will have as low of a memory limit as one can get.

If you have to run `backward` pass as well in production, we give utilities called `checkpoint` and `checkpoint_sequential` that allow you to trade-off memory for speed: https://pytorch.org/docs/stable/checkpoint.html

**Comment by danFromTelAviv** [comment]
Hi @soumith.
I think you missunderstood me. Although that may help a bit. 
What I would like is within each layer for there to be minimal ( or limited ) ram usage as well as in between layers. for example if I have two consecutive conv1d layers of filter len 3 with padding all I need is 5 timesteps in the ram at once to calculate the result for a single final timestep. 

My issues is when I get samples that are 30k timesteps long. Even if I don't keep all of the features in between layers - there are single layers that will not fit in the ram with 30k timesteps.

**Comment by soumith** [comment]
If it spans to a layer-internal granularity, i think such a feature is simply hard to design. It's unlikely that we'll provide it.
You need global limits, and you need every single function in pytorch to be aware of these limits -- it's not a simple change of "oh let's reorder this for-loop". On GPU, it gets even more complex to design a kernel with blocking in mind -- it directly affects speed.

**Comment by danFromTelAviv** [comment]
It would likely be very complex to make this happen ( unless someone comes up with a creative solution ) but it would really help in many production/ low ram situations ( cellphones for example ). I don't think tf has anything like this either so it could be a nice little feature to compete with ; )
I guess it really depends on how many people could find this useful - maybe I'm just looking at it from my narrower view.

**Comment by soumith** [comment]
most production situations that I've heard about or encountered are not low-ram. Even cellphones are getting quite powerful.
The market for low-ram as you mean it is quite low and likely shrinking going forward.

**Comment by danFromTelAviv** [comment]
1-3 gb of ram ( as a single process ) I think will continue to be significant for some time... 
I personally want this for processing [almost] arbitrarily large files - imagine videos that are processed all together, very long audio files, high-resolution images. In our case, I would like to limit my servers to 2 gb of ram due to the costs of a larger server. 
I think the data types will actually grow significantly faster than ram. very large images, videos, 3d data for virtual reality...etc 
I also think that many more processes will want to use DL at the same time so that the competition will intensify for the little ram that is available. Right now very few processes use DL on cellphones. I would be surprised if in the next few years we don't see a sharp increase in such services with specialized chips and that. 
Again - I have not done a study to find the relevance of this feature but it may be worth looking into.
